File: lib-android/src/main/java/com/pranav/lib_android/task/java/ExecuteJavaTask.java
package com.pranav.lib_android.task.java;

import com.pranav.lib_android.interfaces.*;
import com.pranav.lib_android.util.FileUtil;
import dalvik.system.PathClassLoader;
import java.io.*;
import java.security.Permission;
import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.util.concurrent.Executors;
import java.util.concurrent.CountDownLatch;

public class ExecuteJavaTask extends Task {
	
	private final Builder mBuilder;
	private final String clazz;
	public StringBuilder log = new StringBuilder();
	
	public ExecuteJavaTask(Builder builder, String claz) {
		this.mBuilder = builder;
		this.clazz = claz;
	}
	
	@Override
	public String getTaskName() {
		return "Execute java Task";
	}
	
	@Override
	public void doFullTask() throws Exception {
		final PrintStream defaultOut = System.out;
		final PrintStream defaultErr = System.err;
		final String dexFile = FileUtil.getBinDir()
		+ "classes.dex";
		final CountDownLatch latch = new CountDownLatch(1);
		Executors.newSingleThreadExecutor().execute(() -> {
			final OutputStream out = new OutputStream() {
				@Override
				public void write(int b) {
					log.append(String.valueOf((char) b));
				}
				
				@Override
				public String toString() {
					return log.toString();
				}
			};
			System.setOut(new PrintStream(out));
			System.setErr(new PrintStream(out));
			
			PathClassLoader loader = new PathClassLoader(dexFile,
			mBuilder.getClassloader());
			try {

				Class calledClass = loader.loadClass(clazz);
				
				Method method = calledClass.getDeclaredMethod("main", String[].class);
				
				String[] param = {};
				Object result;
				if (Modifier.isStatic(method.getModifiers())) {
					result = method.invoke(null, new Object[] {param});
				} else if (Modifier.isPublic(method.getModifiers())) {
					Object classInstance = calledClass.newInstance();
					result = method.invoke(classInstance, new Object[] {param});
				}
			} catch (Exception e) {
				e.printStackTrace();
			}
			System.setSecurityManager(null);
			latch.countDown();
		});
		try {
			latch.await();
		} catch (InterruptedException ignored) {
		}
		System.setOut(defaultOut);
		System.setErr(defaultErr);
	}
	
	public String getLogs() {
		return log.toString();
	}
}

File: lib-android/src/main/java/com/pranav/lib_android/task/java/ExecuteJavaTask.kt
package com.pranav.lib_android.task.java

import com.pranav.lib_android.interfaces.*
import com.pranav.lib_android.util.FileUtil
import com.pranav.lib_android.util.ConcurrentUtil
import dalvik.system.PathClassLoader
import java.io.PrintStream
import java.io.OutputStream
import java.lang.reflect.Modifier
import java.lang.reflect.InvocationTargetException

class ExecuteJavaTask constructor(
    val mBuilder: Builder,
    val clazz: String
  ): Task() {
	
	var result: Any? = null
	val log = StringBuilder()
	
	override fun getTaskName(): String {
		return "Execute java Task"
	}

	@Throws(InvocationTargetException::class)
	override fun doFullTask() {
		val defaultOut = System.`out`
		val defaultErr = System.err
		val dexFile = FileUtil.getBinDir() + "classes.dex"
		ConcurrentUtil.execute {
			val out = object: OutputStream() {
			  override fun write(b: Int) {
			    log.append(b.toChar())
			  }
			  
			  override fun toString(): String {
			    return log.toString()
			  }
			}
			System.setOut(PrintStream(`out`))
			System.setErr(PrintStream(`out`))
			
			val loader = PathClassLoader(dexFile, mBuilder.getClassloader())
			try {

				val calledClass = loader.loadClass(clazz)
				
				val method = calledClass.getDeclaredMethod("main", Array<String>::class.java)
				
				var param = arrayOf<String>()
				
				if (Modifier.isStatic(method.getModifiers())) {
					val res = method.invoke(null, param as? Any)
					if (res !is Unit) {
					  result = res
					}
				} else if (Modifier.isPublic(method.getModifiers())) {
					val classInstance = calledClass.newInstance()
					val res = method.invoke(classInstance, param as? Any)
					if (res !is Unit) {
					  result = res
					}
				}
				if (result != null) {
				  log.append(result.toString())
				}
			} catch (e: Exception) {
				e.printStackTrace()
			}
			System.setOut(defaultOut)
			System.setErr(defaultErr)
		}
	}
	
	fun getLogs(): String {
		return log.toString()
	}
}